### PR TITLE
[Feat]: support contracts compiled with version 2.1.0

### PR DESCRIPTION
**Pull request by haroune-mohammedi**
Contracts compiled from https://github.com/OpenZeppelin/cairo-contracts commit 83076441b0d7b965f9c533c774f5c3d7b39381cd

With some changes to keep the same function parameters with v0 contracts

```
diff --git a/Scarb.toml b/Scarb.toml
index bf42b24..7694e0d 100644
--- a/Scarb.toml
+++ b/Scarb.toml
@@ -18,4 +18,4 @@ starknet = ">=2.1.0-rc0"
 [[target.starknet-contract]]
 allowed-libfuncs-list.name = "experimental"
 sierra = true
-casm = false
+casm = true
diff --git a/src/token/erc20/erc20.cairo b/src/token/erc20/erc20.cairo
index 9448b8e..e6bc054 100644
--- a/src/token/erc20/erc20.cairo
+++ b/src/token/erc20/erc20.cairo
@@ -45,6 +45,7 @@ mod ERC20 {
         ref self: ContractState,
         name: felt252,
         symbol: felt252,
+        decimals: felt252,
         initial_supply: u256,
         recipient: ContractAddress
     ) {
diff --git a/src/token/erc721/erc721.cairo b/src/token/erc721/erc721.cairo
index 32a265c..e042f31 100644
--- a/src/token/erc721/erc721.cairo
+++ b/src/token/erc721/erc721.cairo
@@ -66,10 +66,10 @@ mod ERC721 {
         name: felt252,
         symbol: felt252,
         recipient: ContractAddress,
-        token_id: u256
+        // token_id: u256
     ) {
         self.initializer(name, symbol);
-        self._mint(recipient, token_id);
+        // self._mint(recipient, token_id);
     }

     //
@@ -364,4 +364,10 @@ mod ERC721 {
             DualCaseSRC5 { contract_address: to }.supports_interface(account::interface::ISRC6_ID)
         }
     }
+
+    #[external(v0)]
+    fn mint(ref self: ContractState, to: ContractAddress, token_id: u256) -> bool {
+        self._mint(to, token_id);
+        true
+    }
 }
 ```

### CODE REVIEW

File: src/actions/shoot.rs [hunk 1]
```diff
@@ -689,43 +724,98 @@ impl GatlingShooter {
...
-                    code: MaybeUnknownErrorCode::Known(StarknetError::ClassAlreadyDeclared),
-                    ..
-                },
-            ))) => {
-                // TODO: get the class_hash from the already declared error
-                warn!("Contract already declared class_hash={:#064x}", class_hash);
-                Ok(class_hash)
+            Err(e) => {
+                // TODO: why panic here ?
+                panic!("Could not declare contract: {e}");
```
Comment (d-roak) [inline src/actions/shoot.rs:754]: shouldn't this be removed?

File: src/actions/shoot.rs [hunk 2]
```diff
@@ -689,43 +724,98 @@ impl GatlingShooter {
...
+                info!(
+                    "Contract declared successfully at {:#064x}",
+                    tx_resp.class_hash
+                );
+                self.nonce += felt!("1");
+                Ok(tx_resp.class_hash)
             }
             Err(e) => {
+                // TODO: why panic here ?
                 panic!("Could not declare contract: {e}");
```
Comment (d-roak) [inline src/actions/shoot.rs:802]: Same as the other one

File: src/actions/shoot.rs (hunk 1 above)
Comment (haroune-mohammedi) [reply src/actions/shoot.rs]: Done

File: src/actions/shoot.rs (hunk 2 above)
Comment (haroune-mohammedi) [reply src/actions/shoot.rs]: Done

File: src/actions/shoot.rs [hunk 3]
```diff
@@ -689,40 +721,83 @@ impl GatlingShooter {
...
+        if self.check_already_declared(class_hash).await? {
+            return Ok(class_hash);
+        }
+
+        // `SingleOwnerAccount` defaults to checking nonce and estimating fees against the latest
+        // block. Optionally change the target block to pending with the following line:
+        self.account.set_block_id(BlockId::Tag(BlockTag::Pending));
+
+        // We need to flatten the ABI into a string first
+        let flattened_class = contract_artifact.flatten().unwrap();
```
Comment (tdelabro) [inline src/actions/shoot.rs:773]: Explain why it is ok to unwrap in a comment

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: src/actions/shoot.rs [hunk 4]
```diff
@@ -78,6 +80,8 @@ impl GatlingShooter {
             signer.clone(),
             config.deployer.address,
             chain_id::TESTNET,
+            // TODO: How should we decide what to use here ?
+            ExecutionEncoding::Legacy,
```
[the remark this answers is not in the document]
Comment (haroune-mohammedi) [reply src/actions/shoot.rs]: Updated comment

File: src/actions/shoot.rs (hunk 3 above)
Comment (haroune-mohammedi) [reply src/actions/shoot.rs]: Replaced with ? operator

File: src/actions/shoot.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tdelabro) [reply src/actions/shoot.rs]: But you also have to make it configurable. So one or the other is used depending of the type of contracts you want to deploy. No?

**Review by tdelabro — APPROVED** [review]

**Pull request merged**